src/content/docs/build/guides/build-e2e-dapp.mdx

```diff
@@ -24,33 +26,72 @@ After meeting the [prerequisites](#prerequisites) and [getting set up](#setup) a
 You must have:
 
 - [node and npm](https://nodejs.org/en/)
+- [VSCode IDE](https://code.visualstudio.com/) or [Cursor](https://www.cursor.com/)
 
 Although we will explain some React decisions, we are not going to deep dive into how React works; so we assume you have some previous experience with React.
 
 ## Setup
 
-In this section, we will create a `my-first-dapp` directory to hold our project files, both client-side code (React based) and the Move code (our smart contract).
+In this section, we will setup three things:
+
+1. API key using [Geomi](https://geomi.dev/)
+2. Project using [create-aptos-dapp](/build/create-aptos-dapp) which will hold our project files, both client-side code (React based) and the Move code (our smart contract)
+3. [Move On Aptos](/build/smart-contracts/move-vscode-extension) VSCode extension
+
+First let's create an API key using [Geomi](https://geomi.dev/). The Aptos TypeScript SDK uses Geomi API by default, and without an API key, you will get very low rate limits.
+
+<Steps>
+  1. **Visit Geomi**: Go to [geomi.dev](https://geomi.dev/) in your web browser.
+
+  2. **Sign up or Log in**: Create a new account or log in to your existing Geomi account.
+
+  3. **Create a new project**: Click on "Create New Project" to create a new project and name it `my-first-dapp`.
+     ![Geomi API key creation 1](~/images/screenshots/geomi-api-1.png)
+     ![Geomi API key creation 2](~/images/screenshots/geomi-api-2.png)
+
+  4. **Navigate to API Resource tab**: Navigate to the API Resource tab in your dashboard.
+     ![Geomi API key creation 3](~/images/screenshots/geomi-api-3.png)
+
+  5. **Create New API Resource**: Name it `my-first-dapp-api` and set the network to `Devnet`.
+     ![Geomi API key creation 4](~/images/screenshots/geomi-api-4.png)
+
+  6. **Configure API Key**: Since we will be using the API key in our web app, toggle on `Client usage`. Here you can configure allowed URLs and extension IDs that can use the API key. Set `Allowed URLs` to `http://localhost:5173/`, which is the development server we will be using. Also set `Per IP rate limit` to `10,000`
```

Comment:
The port actually doesn't matter, we only look at the host: https://github.com/aptos-labs/api-gateway/blob/a11ee3abffee764b3e9567c5ce6c06dd1fe19479/src/http_proxy/http_handler.rs#L225.